### ISSUE TITLE
Deprecated watch/watchlist API calls misreported under get/list call metrics

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#104439]
While investigating a recent issue on a v1.20 cluster, I found out that if clients are using the old [WATCH](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L906-L907) or [WATCHLIST](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L929-L930) apis (deprecated in k8s v1.11, but still supported by apiserver) those calls are actually being counted under GET and LIST call metrics respectively. This is incorrect behavior and pollutes the latency metrics for GET/LIST calls.

This issue can easily be reproduced as follows:
- Issue the following API calls:
```
# watchlist call
kubectl get --raw /api/v1/watch/pods?resourceVersion=100

# watch call
kubectl get --raw /api/v1/watch/namespaces/kube-system/pods/somepod?resourceVersion=100
```
- You can even omit the RV parameter. The behavior in that case is that all existing objects come through an initial wave of ADD-type watch events (instead of the current list+watch pattern where an initial list call gets all the objects and then we begin taking watch events)
- While you're making those calls, observe how GET/LIST-related metrics go up:
```
$ while true; do
   kubectl get --raw /metrics | grep "apiserver_request_duration_seconds_count" | grep "\"pods\"" | grep "\"LIST\"" | grep "\"cluster\"" | awk '{print $2}'
   sleep 1
done

61076
61076
61080
61086
61092
61092
```
```
$ while true; do
   kubectl get --raw /metrics | grep "apiserver_request_duration_seconds_count" | grep "\"pods\"" | grep "\"GET\"" | grep "subresource=\"\"" | awk '{print $2}'
   sleep 1
done

5.537744e+06
5.537745e+06
5.537751e+06
5.537758e+06
5.537764e+06
5.537764e+06
5.537764e+06
```

/sig api-machinery
/sig instrumentation
/kind bug

**Comment by shyamjvs** [comment]
Digging into the code a bit:
- the metric is being recorded [here in MonitorRequest()](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L420)
- where the reportedVerb is being computed like this:
```
reportedVerb := cleanVerb(canonicalVerb(strings.ToUpper(req.Method), scope), req)
```
- and [canonicalVerb()](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L508) has below logic, so it should return either GET or LIST (since the input request's HTTP verb is GET):
```
func canonicalVerb(verb string, scope string) string {
	switch verb {
	case "GET", "HEAD":
		if scope != "resource" && scope != "" {
			return "LIST"
		}
		return "GET"
	default:
		return verb
	}
}
```
- So we call [cleanVerb()](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L520) with either GET or LIST. And in either cases it's logic doesn't change the verb:
```
func cleanVerb(verb string, request *http.Request) string {
	reportedVerb := verb
	if verb == "LIST" {
		// see apimachinery/pkg/runtime/conversion.go Convert_Slice_string_To_bool
		if values := request.URL.Query()["watch"]; len(values) > 0 {
			if value := strings.ToLower(values[0]); value != "0" && value != "false" {
				reportedVerb = "WATCH"
			}
		}
	}
	// normalize the legacy WATCHLIST to WATCH to ensure users aren't surprised by metrics
	if verb == "WATCHLIST" {
		reportedVerb = "WATCH"
	}
	if verb == "PATCH" && request.Header.Get("Content-Type") == string(types.ApplyPatchType) && utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply) {
		reportedVerb = "APPLY"
	}
	if validRequestMethods.Has(reportedVerb) {
		return reportedVerb
	}
	return OtherRequestMethod
}
```

So the fix essentially is we shouldn't be ignoring the verb [coming to MonitorRequest()](https://github.com/kubernetes/kubernetes/blob/v1.20.7/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L451) from InstrumentRouteFunc() as only that carries the context that these are deprecated WATCH/WATCHLIST calls.

**Comment by shyamjvs** [comment]
Aah! Just as I was about to make a fix, saw that @wojtek-t fixed this exact bug couple weeks ago :) https://github.com/kubernetes/kubernetes/pull/104161

Since cherry-picks have also been merged for it up to 1.20, I'm closing this issue. Sorry for the noise.
